### PR TITLE
Transfer LPs consolidation

### PR DESCRIPTION
**Pull request by grandizzy**
- addresses https://github.com/ajna-finance/contracts/issues/620 and https://github.com/ajna-finance/contracts/issues/621 by introducing approved transferors
- changed `approveLpOwnership` to allow array of indexes and amounts as arguments (reduce number of tx / gas and number of external calls from `PositionManager`)
- introduced `approvedTransferors` mapping which hold addresses allowed to transfer LPs
- introduced `Pool.approveLpTransferors` and `Pool.revokeLpTransferors` lender external function to allow addresses as LPs transferors
- on `Lender.transferLPs`, check if `msg.sender` is an allowed transferor (only when is not the new owner). If not then reverts with `TransferorNotApproved` error
- position manager should be approved as transferor by lenders memorializing their positions
- added `ApproveLpOwnership`, `ApproveLpTransferors` and `RevokeLpTransferors` events
- fixed tests, added new

### CODE REVIEW

File: src/base/Pool.sol [hunk 1]
```diff
@@ -165,11 +167,54 @@ abstract contract Pool is Clone, ReentrancyGuard, Multicall, IPool {
...
-        address newOwner,
-        uint256 index_,
-        uint256 amount_
-    ) external nonReentrant {
-        _lpAllowances[msg.sender][newOwner][index_] = amount_;
+        address newOwner_,
+        uint256[] calldata indexes_,
+        uint256[] calldata amounts_
+    ) external override nonReentrant {
+        mapping(uint256 => uint256) storage allowances = _lpAllowances[msg.sender][newOwner_];
```
Comment (MikeHathaway) [inline src/base/Pool.sol:174]: Is the local mapping declaration necessary for gas savings?

**Review by MikeHathaway — APPROVED** [review]
LGTM - still have some uncertainties about how this change interacts with the `depositTime` issue

**Comment by grandizzy** [comment]
> LGTM - still have some uncertainties about how this change interacts with the `depositTime` issue

Well it doesn't 😀 it only address other lp transfer issues. I will add also events for these new actions

**Comment by grandizzy** [comment]
> > LGTM - still have some uncertainties about how this change interacts with the `depositTime` issue
> 
> Well it doesn't grinning it only address other lp transfer issues. I will add also events for these new actions

Events added with https://github.com/ajna-finance/contracts/pull/625/commits/a9678cdb24c05c40af6d2359508492f650fabfe9

**Review by EdNoepel — APPROVED** [review]
LGTM

File: src/base/Pool.sol (hunk 1 above)
Comment (grandizzy) [reply src/base/Pool.sol:174]: yes, that's for gas savings purposes

**Review by dmitriia — COMMENTED** [review]
Looks ok, but the allowance approach used doesn't protect from race condition.

If Bob gave Alice allowance of 10, and then wants to give her 11, she still can front run this call and spend fully this existing 10, so `allowances[index] == 0` be true and Bob will sucessfully set 11, so Alice can end up spending 21 in total, which not what Bob intended.

The proper way looks to be going increase/decrease route, something like:

https://github.com/ajna-finance/contracts/blob/b00e3e51d04acf7ef5bde168c6871a486b4e75a3/src/base/Pool.sol#L170-L190

```solidity
-   function approveLpOwnership(
+   function increaseLpAllowance(
        address newOwner_,
        uint256[] calldata indexes_,
        uint256[] calldata amounts_
    ) external override nonReentrant {
        mapping(uint256 => uint256) storage allowances = _lpAllowances[msg.sender][newOwner_];

        uint256 indexesLength = indexes_.length;
        uint256 index;

        for (uint256 i = 0; i < indexesLength; ) {
            index = indexes_[i];

-           // revert if allowance at index is already set (not 0) and the new allowance does not reset the old one (not 0)
-           // this prevents possible attack where LPs receiver (newOwner) frontruns owner allowance calls to transfer more than allowed
-           if (allowances[index] != 0 && amounts_[i] != 0) revert AllowanceAlreadySet();

-           allowances[index] = amounts_[i];
+           allowances[index] += amounts_[i];

            unchecked { ++i; }
        }
```

The `if (allowances[index] != 0 && amounts_[i] != 0` check can also cause issues when say Bob gave Alice allowance of 10, which she didn't immediately used, Bob's shares dropped to 9, then she used LP transfer, and have residual allowance of 1 as a result.

Next time if Bob will want to give her allowance he will not be able to do it directly, Alice would need to spend the residual first.

I see increase / decrease / revoke (set all to 0) without additional conditions as a good enough approach here.

**Comment by dmitriia** [comment]
Also, comments can be updated:

https://github.com/ajna-finance/contracts/blob/b00e3e51d04acf7ef5bde168c6871a486b4e75a3/src/interfaces/pool/commons/IPoolEvents.sol#L237-L245

```diff
    /**
     *  @notice Emitted when lender removes addresses from the LPs transferors whitelist.
-    *  @param  lender      Recipient that approves new owner for LPs.
+    *  @param  lender      Recipient that revokes the transferors.
     *  @param  transferors List of addresses that won't be able to transfer LPs to lender anymore.
     */
    event RevokeLpTransferors(
        address indexed lender,
        address[] transferors
    );
```

https://github.com/ajna-finance/contracts/blob/b00e3e51d04acf7ef5bde168c6871a486b4e75a3/src/interfaces/pool/commons/IPoolEvents.sol#L39-L47

```diff
    /**
     *  @notice Emitted when lender whitelists addresses to accept LPs from.
-    *  @param  lender      Recipient that approves new owner for LPs.
+    *  @param  lender      Recipient that approves the transferors.
     *  @param  transferors List of addresses that can transfer LPs to lender.
     */
    event ApproveLpTransferors(
        address indexed lender,
        address[] transferors
    );
```

**Comment by grandizzy** [comment]
> Also, comments can be updated:
> 
> https://github.com/ajna-finance/contracts/blob/b00e3e51d04acf7ef5bde168c6871a486b4e75a3/src/interfaces/pool/commons/IPoolEvents.sol#L237-L245
> 
> ```diff
>     /**
>      *  @notice Emitted when lender removes addresses from the LPs transferors whitelist.
> -    *  @param  lender      Recipient that approves new owner for LPs.
> +    *  @param  lender      Recipient that revokes the transferors.
>      *  @param  transferors List of addresses that won't be able to transfer LPs to lender anymore.
>      */
>     event RevokeLpTransferors(
>         address indexed lender,
>         address[] transferors
>     );
> ```
> 
> https://github.com/ajna-finance/contracts/blob/b00e3e51d04acf7ef5bde168c6871a486b4e75a3/src/interfaces/pool/commons/IPoolEvents.sol#L39-L47
> 
> ```diff
>     /**
>      *  @notice Emitted when lender whitelists addresses to accept LPs from.
> -    *  @param  lender      Recipient that approves new owner for LPs.
> +    *  @param  lender      Recipient that approves the transferors.
>      *  @param  transferors List of addresses that can transfer LPs to lender.
>      */
>     event ApproveLpTransferors(
>         address indexed lender,
>         address[] transferors
>     );
> ```

we introduced the increase/decrease/revoke functionality with https://github.com/ajna-finance/contracts/commit/6517aee1db3224f7a14ce775defc6f20af29d076

**Pull request merged**